### ISSUE TITLE
`@cached_property` rejected on method with enclosing-class `TypeVar` in return type

### ISSUE THREAD

**Issue by dsfaccini** [astral-sh/ty#3256]
### Summary
ty rejects `@cached_property` on a method whose return type references an enclosing-class `TypeVar`.

Playground link: https://play.ty.dev/13c53f50-5186-4204-a038-43f54724e78f

### Repro (no external deps)
`ty_repro_cached_property_generic.py`:
```python
from functools import cached_property
from typing import Generic, TypeVar

T = TypeVar('T')


class Container(Generic[T]):
    @cached_property
    def value(self) -> list[T] | None:
        return None
```

Run:
```bash
uvx ty check ty_repro_cached_property_generic.py
```

### Actual result
```
error[invalid-argument-type]: Argument to bound method `__init__` is incorrect
  --> ty_repro_cached_property_generic.py:8:5
   |
 7 | class Container(Generic[T]):
 8 |     @cached_property
   |     ^^^^^^^^^^^^^^^^ Expected `(Any, /) -> None`, found `def value(self) -> list[T@Container] | None`
 9 |     def value(self) -> list[T] | None:
10 |         return None
   |
info: Method defined here
   --> stdlib/functools.pyi:418:9
    |
418 |     def __init__(self, func: Callable[[Any], _T_co]) -> None: ...
    |         ^^^^^^^^       ---------------------------- Parameter declared here
    |
```

### Expected result
No diagnostic. The decorated method satisfies `Callable[[Any], list[T] | None]`, which is compatible with `cached_property.__init__`'s `Callable[[Any], _T_co]`. A non-generic return type (e.g. `list[int]` or `list[str]`) on the same class does not trigger the error — only a return type that mentions the enclosing-class `TypeVar` does. Mypy and pyright both accept this.

### Notes
ty 0.0.29 (438a78d68 2026-04-05) via `uvx ty check`. The closed #1446 added `cached_property` support but discussion only covered non-generic classes; the fix does not handle the case where the method's return type references an enclosing-class `TypeVar`. Closest neighbours: #2520 (generic decorator on method annotated with `Self`) and #1840 (user-defined generic data descriptors).

**Comment by carljm** [comment]
I suspect this is due to the old constraint solver not matching up the typevar in the return annotation of the method with the generic definition of `cached_property` correctly.

**Comment by carljm** [comment]
My suspicion wasn't quite right. #3670 is a duplicate that has a more accurate analysis of the root cause and fix.
